verbose output confusing, as it shows test passed

The example given in the readme is kinda weird when passed the `-v` flag.

```
(subtest) $ pytest -v test_sub_pytest.py  --tb=no
================= test session starts =================
platform darwin -- Python 3.8.0a1, pytest-4.4.0, py-1.8.0, pluggy-0.9.0 -- 
plugins: subtests-0.1.0
collected 1 item                                      

test_sub_pytest.py::test 
test_sub_pytest.py::test PASSED                 [100%]

========= 2 failed, 1 passed in 0.05 seconds ==========
```

The double `::test` is a bit odd.
As is the `PASSED`.
And, well, with 1 test that has 5 subtests, where does the `2 failed 1 passed` come from?
Specifically, the `1 passed`?

Hi @okken and @jurisbu,

> The double ::test is a bit odd.

Strange, I get different results here:

```
λ pytest .tmp\test_ut.py -v --tb=no
======================== test session starts ========================
platform win32 -- Python 3.7.1, pytest-4.4.0, py-1.8.0, pluggy-0.9.0 -- ...
plugins: subtests-0.1.1.dev4+gea4c7be
collected 1 item

.tmp/test_ut.py::test PASSED                                   [100%]
.tmp/test_ut.py::test FAILED                                   [100%]
.tmp/test_ut.py::test PASSED                                   [100%]
.tmp/test_ut.py::test FAILED                                   [100%]
.tmp/test_ut.py::test PASSED                                   [100%]
.tmp/test_ut.py::test PASSED                                   [100%]

================ 2 failed, 1 passed in 0.04 seconds =================
```

The `100%` bit is indeed a bit weird, perhaps we should not show the percentage at all for subtests? 

```
.tmp/test_ut.py::test PASSED                                   
.tmp/test_ut.py::test FAILED                                   
.tmp/test_ut.py::test PASSED                                   
.tmp/test_ut.py::test FAILED                                   
.tmp/test_ut.py::test PASSED                                   
.tmp/test_ut.py::test PASSED                                   [100%]
```

e.g, only show the percentage when we reach the end of the test.

> And, well, with 1 test that has 5 subtests, where does the 2 failed 1 passed come from?
Specifically, the 1 passed?

I've tried to mimic the output from `unittest` when running subtests:

```python
import unittest

class T(unittest.TestCase):

    def test_foo(self):
        for i in range(5):
            with self.subTest("custom message", i=i):
                self.assertEqual(i % 2, 0)

if __name__ == "__main__":
    unittest.main()
```

```
λ python .tmp\test_ut.py

======================================================================
FAIL: test_foo (__main__.T) [custom message] (i=1)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".tmp\test_ut.py", line 10, in test_foo
    self.assertEqual(i % 2, 0)
AssertionError: 1 != 0

======================================================================
FAIL: test_foo (__main__.T) [custom message] (i=3)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".tmp\test_ut.py", line 10, in test_foo
    self.assertEqual(i % 2, 0)
AssertionError: 1 != 0

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (failures=2)
```

It says above that it executed 1 test, but reported 2 failures. We actually have 5 sub-tests, with 3 of them passing and 2 failures, but the passing sub-tests don't count towards the summary. I followed the same approach to reporting for the `subtests` fixture, even though I agree it is a bit odd.

Of course we can change the output when running unittest subtests under pytest, or perhaps adopt a different reporting strategy when using `subtests` fixture?

Sorry for the delay, Sounds like it's understood now.
It is NOT ok for a test to pass if any of the subtests fail.

The unittest output is confusing, yes. 
However, it does NOT claim that a test passed if a subtest failed.
It doesn't say it failed either, which is definitely weird, but better than saying it passed.
Actually, I think those who implemented subTest in unittest just forgot to decide on an outcome.
It doesn't even use a newline where a result decision should be.

So my opinion should be obvious by now. 
Any failure in a test should fail the test.
At the very least, PASSED is wrong.
The notion that PASSED just means there were no asserts outside of the subtests is a very hard thing to explain.

With this:
```python
import unittest

class T(unittest.TestCase):
    def test_foo(self):
        for i in range(3):
            with self.subTest("custom message", i=i):
                self.assertEqual(i % 2, 0)

    def test_pass(self):
        pass

    def test_fail(self):
        assert False, 'fail intentionally'
```

```bash
(pytest_wlan) c:\sub>python -m unittest -v test_sub.py
test_fail (test_sub.T) ... FAIL
test_foo (test_sub.T) ... test_pass (test_sub.T) ... ok

======================================================================
FAIL: test_fail (test_sub.T)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\sub\test_sub.py", line 13, in test_fail
    assert False, 'fail intentionally'
AssertionError: fail intentionally

======================================================================
FAIL: test_foo (test_sub.T) [custom message] (i=1)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\sub\test_sub.py", line 7, in test_foo
    self.assertEqual(i % 2, 0)
AssertionError: 1 != 0

----------------------------------------------------------------------
Ran 3 tests in 0.002s

FAILED (failures=2)
```

with pytest:
```
c:\sub>python -m pytest -v test_sub.py
=============================== test session starts ===============================
platform win32 -- Python 3.7.1, pytest-4.4.0, py-1.7.0, pluggy-0.9.0 -- C:\Users\okken\projects\pytest_wlan\venv\Scripts\python.exe
cachedir: .pytest_cache
metadata: {'Python': '3.7.1', 'Platform': 'Windows-10-10.0.17134-SP0', 'Packages': {'pytest': '4.4.0', 'py': '1.7.0', 'pluggy': '0.9.0'}, 'Plugins': {'subtests': '0.2.0''}}
rootdir: c:\sub
plugins: subtests-0.2.0
collected 3 items

test_sub.py::T::test_fail FAILED                                             [ 33%]
test_sub.py::T::test_foo FAILED                                          [ 66%]
test_sub.py::T::test_foo PASSED                                              [ 66%]
test_sub.py::T::test_pass PASSED                                             [100%]

==================================== FAILURES =====================================
___________________________________ T.test_fail ___________________________________

self = <test_sub.T testMethod=test_fail>

    def test_fail(self):
>       assert False, 'fail intentionally'
E       AssertionError: fail intentionally
E       assert False

test_sub.py:13: AssertionError
________________________ T.test_foo [custom message] (i=1) ________________________

self = <test_sub.T testMethod=test_foo>

    def test_foo(self):
        for i in range(3):
            with self.subTest("custom message", i=i):
>               self.assertEqual(i % 2, 0)
E               AssertionError: 1 != 0

test_sub.py:7: AssertionError
======================= 2 failed, 2 passed in 0.07 seconds ========================

```

One more test case that is interesting.
This file:
```python
import unittest

class T(unittest.TestCase):
    def test_fail(self):
        with self.subTest():
           self.assertEqual(1, 2)
```

No passing subtests, just one failure. Still shows PASSED.
I would expect FAILED and the bottom line to show `1 failed`
Maybe pytest itself is calling `pytest_runtest_logreport()` at the end and causing an extra test to be counted.

```
pytest -v test_sub.py
=============================== test session starts ===============================
platform win32 -- Python 3.7.1, pytest-4.4.0, ...
plugins: subtests-0.2.0
collected 1 item

test_sub.py::T::test_fail FAILED                                         [100%]
test_sub.py::T::test_fail PASSED                                             [100%]

==================================== FAILURES =====================================
_____________________________ T.test_fail (<subtest>) _____________________________

self = <test_sub.T testMethod=test_fail>

    def test_fail(self):
        with self.subTest():
>          self.assertEqual(1, 2)
E          AssertionError: 1 != 2

test_sub.py:6: AssertionError
======================= 1 failed, 1 passed in 0.05 seconds ========================
```

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -1,6 +1,13 @@
 CHANGELOG
 =========
 
+0.2.1 (2019-04-04)
+------------------
+
+* Fix verbose output reporting on Linux (`#7`_).
+
+.. _#7: https://github.com/pytest-dev/pytest-subtests/issues/7
+
 0.2.0 (2019-04-03)
 ------------------
 

--- a/README.rst
+++ b/README.rst
@@ -23,6 +23,9 @@ unittest ``subTest()`` support and ``subtests`` fixture.
     :target: https://dev.azure.com/pytest-dev/pytest-subtests
     :alt: See Build Status on Azure
 
+.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
+    :target: https://github.com/ambv/black
+
 ----
 
 This `pytest`_ plugin was generated with `Cookiecutter`_ along with `@hackebrot`_'s `cookiecutter-pytest-plugin`_ template.

--- a/pytest_subtests.py
+++ b/pytest_subtests.py
@@ -1,3 +1,4 @@
+import sys
 from contextlib import contextmanager
 from time import time
 
@@ -8,6 +9,16 @@ from _pytest.outcomes import OutcomeException
 from _pytest.reports import TestReport
 from _pytest.runner import CallInfo
 from _pytest.unittest import TestCaseFunction
+
+if sys.version_info[:2] < (3, 7):
+
+    @contextmanager
+    def nullcontext():
+        yield
+
+
+else:
+    from contextlib import nullcontext
 
 
 @attr.s
@@ -80,13 +91,19 @@ def pytest_unconfigure():
 
 @pytest.fixture
 def subtests(request):
-    yield SubTests(request.node.ihook, request.node)
+    capmam = request.node.config.pluginmanager.get_plugin("capturemanager")
+    if capmam is not None:
+        suspend_capture_ctx = capmam.global_and_fixture_disabled
+    else:
+        suspend_capture_ctx = nullcontext
+    yield SubTests(request.node.ihook, request.node, suspend_capture_ctx)
 
 
 @attr.s
 class SubTests(object):
     ihook = attr.ib()
     item = attr.ib()
+    suspend_capture_ctx = attr.ib()
 
     @contextmanager
     def test(self, msg=None, **kwargs):
@@ -100,7 +117,8 @@ class SubTests(object):
         call_info = CallInfo(None, exc_info, start, stop, when="call")
         sub_report = SubTestReport.from_item_and_call(item=self.item, call=call_info)
         sub_report.context = SubTestContext(msg, kwargs.copy())
-        self.ihook.pytest_runtest_logreport(report=sub_report)
+        with self.suspend_capture_ctx():
+            self.ihook.pytest_runtest_logreport(report=sub_report)
 
 
 def pytest_report_to_serializable(report):

--- a/tests/test_subtests.py
+++ b/tests/test_subtests.py
@@ -1,13 +1,16 @@
+import sys
+
 import pytest
 
 
 @pytest.mark.parametrize("mode", ["normal", "xdist"])
 class TestFixture:
     """
-    TODO: test skips, xfails
+    Tests for ``subtests`` fixture.
     """
 
-    def test_simple_terminal_out(self, testdir, mode):
+    @pytest.fixture
+    def simple_script(self, testdir):
         testdir.makepyfile(
             """
             def test_foo(subtests):
@@ -16,6 +19,8 @@ class TestFixture:
                         assert i % 2 == 0
         """
         )
+
+    def test_simple_terminal_normal(self, simple_script, testdir, mode):
         if mode == "normal":
             result = testdir.runpytest()
             expected_lines = ["collected 1 item"]
@@ -23,6 +28,38 @@ class TestFixture:
             pytest.importorskip("xdist")
             result = testdir.runpytest("-n1")
             expected_lines = ["gw0 [1]"]
+
+        expected_lines += [
+            "* test_foo [[]custom[]] (i=1) *",
+            "* test_foo [[]custom[]] (i=3) *",
+            "* 2 failed, 1 passed in *",
+        ]
+        result.stdout.fnmatch_lines(expected_lines)
+
+    def test_simple_terminal_verbose(self, simple_script, testdir, mode):
+        if mode == "normal":
+            result = testdir.runpytest("-v")
+            expected_lines = [
+                "*collected 1 item",
+                "test_simple_terminal_verbose.py::test_foo PASSED *100%*",
+                "test_simple_terminal_verbose.py::test_foo FAILED *100%*",
+                "test_simple_terminal_verbose.py::test_foo PASSED *100%*",
+                "test_simple_terminal_verbose.py::test_foo FAILED *100%*",
+                "test_simple_terminal_verbose.py::test_foo PASSED *100%*",
+                "test_simple_terminal_verbose.py::test_foo PASSED *100%*",
+            ]
+        else:
+            pytest.importorskip("xdist")
+            result = testdir.runpytest("-n1", "-v")
+            expected_lines = [
+                "gw0 [1]",
+                "*gw0*100%* test_simple_terminal_verbose.py::test_foo*",
+                "*gw0*100%* test_simple_terminal_verbose.py::test_foo*",
+                "*gw0*100%* test_simple_terminal_verbose.py::test_foo*",
+                "*gw0*100%* test_simple_terminal_verbose.py::test_foo*",
+                "*gw0*100%* test_simple_terminal_verbose.py::test_foo*",
+                "*gw0*100%* test_simple_terminal_verbose.py::test_foo*",
+            ]
 
         expected_lines += [
             "* test_foo [[]custom[]] (i=1) *",
@@ -55,12 +92,12 @@ class TestFixture:
 
 class TestSubTest:
     """
-    # TODO: test skips, xfails
+    Test Test.subTest functionality.
     """
 
-    @pytest.mark.parametrize("runner", ["unittest", "pytest-normal", "pytest-xdist"])
-    def test_simple_terminal_out(self, testdir, runner):
-        p = testdir.makepyfile(
+    @pytest.fixture
+    def simple_script(self, testdir):
+        return testdir.makepyfile(
             """
             from unittest import TestCase, main
 
@@ -75,8 +112,12 @@ class TestSubTest:
                 main()
         """
         )
+
+    @pytest.mark.parametrize("runner", ["unittest", "pytest-normal", "pytest-xdist"])
+    def test_simple_terminal_normal(self, simple_script, testdir, runner):
+
         if runner == "unittest":
-            result = testdir.runpython(p)
+            result = testdir.run(sys.executable, simple_script)
             result.stderr.fnmatch_lines(
                 [
                     "FAIL: test_foo (__main__.T) [custom] (i=1)",
@@ -89,12 +130,57 @@ class TestSubTest:
             )
         else:
             if runner == "pytest-normal":
-                result = testdir.runpytest(p)
+                result = testdir.runpytest(simple_script)
                 expected_lines = ["collected 1 item"]
             else:
                 pytest.importorskip("xdist")
-                result = testdir.runpytest("-n1")
+                result = testdir.runpytest(simple_script, "-n1")
                 expected_lines = ["gw0 [1]"]
+            result.stdout.fnmatch_lines(
+                expected_lines
+                + [
+                    "* T.test_foo [[]custom[]] (i=1) *",
+                    "E  * AssertionError: 1 != 0",
+                    "* T.test_foo [[]custom[]] (i=3) *",
+                    "E  * AssertionError: 1 != 0",
+                    "* 2 failed, 1 passed in *",
+                ]
+            )
+
+    @pytest.mark.parametrize("runner", ["unittest", "pytest-normal", "pytest-xdist"])
+    def test_simple_terminal_verbose(self, simple_script, testdir, runner):
+
+        if runner == "unittest":
+            result = testdir.run(sys.executable, simple_script, "-v")
+            result.stderr.fnmatch_lines(
+                [
+                    "test_foo (__main__.T) ... ",
+                    "FAIL: test_foo (__main__.T) [custom] (i=1)",
+                    "AssertionError: 1 != 0",
+                    "FAIL: test_foo (__main__.T) [custom] (i=3)",
+                    "AssertionError: 1 != 0",
+                    "Ran 1 test in *",
+                    "FAILED (failures=2)",
+                ]
+            )
+        else:
+            if runner == "pytest-normal":
+                result = testdir.runpytest(simple_script, "-v")
+                expected_lines = [
+                    "*collected 1 item",
+                    "test_simple_terminal_verbose.py::T::test_foo FAILED *100%*",
+                    "test_simple_terminal_verbose.py::T::test_foo FAILED *100%*",
+                    "test_simple_terminal_verbose.py::T::test_foo PASSED *100%*",
+                ]
+            else:
+                pytest.importorskip("xdist")
+                result = testdir.runpytest(simple_script, "-n1", "-v")
+                expected_lines = [
+                    "gw0 [1]",
+                    "*gw0*100%* FAILED test_simple_terminal_verbose.py::T::test_foo*",
+                    "*gw0*100%* FAILED test_simple_terminal_verbose.py::T::test_foo*",
+                    "*gw0*100%* PASSED test_simple_terminal_verbose.py::T::test_foo*",
+                ]
             result.stdout.fnmatch_lines(
                 expected_lines
                 + [

